lxd/main_forkproxy.go

```diff
@@ -5,129 +5,281 @@ import (
+		ret = dup3(sk_fds[1], FORKPROXY_UDS_SOCK_FD_NUM, O_CLOEXEC);
+		if (ret < 0) {
+			fprintf(stderr,
+				"%s - Failed to duplicate fd %d to fd 200\n",
+				strerror(errno), sk_fds[1]);
 			_exit(1);
 		}
 
-		childPid = fork();
-		if (childPid < 0) {
-			fprintf(stderr, "Failed to fork proxy daemon: %s\n", strerror(errno));
+		ret = close(sk_fds[1]);
+		if (ret < 0) {
+			fprintf(stderr, "%s - Failed to close socket fd %d\n",
+				strerror(errno), sk_fds[1]);
 			_exit(1);
-		} else if (childPid != 0) {
-			fprintf(pidFile, "%d", childPid);
-			fclose(pidFile);
-			fclose(stdin);
-			fclose(stdout);
-			fclose(stderr);
-			_exit(0);
-		} else {
-			ret = setsid();
-			if (ret < 0) {
-				fprintf(stderr, "Failed to setsid in proxy daemon: %s\n", strerror(errno));
-				_exit(1);
-			}
 		}
-	}
+	} else {
+		whoami = FORKPROXY_PARENT;
 
-	// Cannot pass through -1 to runCommand since it is interpreted as a flag
-	fdnum = fdnum == 0 ? -1 : fdnum;
+		close(sk_fds[1]);
 
-	ret = snprintf(fdpath, sizeof(fdpath), "/proc/self/fd/%d", fdnum);
-	if (ret < 0 || (size_t)ret >= sizeof(fdpath)) {
-		fprintf(stderr, "Failed to format file descriptor path\n");
-		_exit(1);
-	}
+		// Attach to the user namespace of the listener
+		attach_userns(connect_pid);
 
-	// Join the listener ns if not already setup
-	if (access(fdpath, F_OK) < 0) {
 		// Attach to the network namespace of the listener
-		if (dosetns(listen_pid, "net") < 0) {
-			fprintf(stderr, "Failed setns to listener network namespace: %s\n", strerror(errno));
+		ret = dosetns(connect_pid, "net");
+		if (ret < 0) {
+			fprintf(stderr, "Failed setns to listener network namespace: %s\n",
+				strerror(errno));
+			_exit(EXIT_FAILURE);
+		}
+
+		// Attach to the mount namespace of the listener
+		ret = dosetns(connect_pid, "mnt");
+		if (ret < 0) {
+			fprintf(stderr, "Failed setns to listener mount namespace: %s\n",
+				strerror(errno));
+			_exit(EXIT_FAILURE);
+		}
+
+		if (strncmp(connect_addr, "unix:", sizeof("unix:") - 1) == 0) {
+			cur = connect_addr + sizeof("unix:") - 1;
+			if (*cur != '@')
+				if (unlink(cur) < 0 && errno != ENOENT)
+					fprintf(stderr, "%s - Failed to remove old connection socket \"%s\"",
+						strerror(errno), cur);
+		}
+
+		ret = dup3(sk_fds[0], FORKPROXY_UDS_SOCK_FD_NUM, O_CLOEXEC);
+		if (ret < 0) {
+			fprintf(stderr,
+				"%s - Failed to duplicate fd %d to fd 200\n",
+				strerror(errno), sk_fds[1]);
 			_exit(1);
 		}
-	} else {
-		// Join the connector ns now
-		if (dosetns(connect_pid, "net") < 0) {
-			fprintf(stderr, "Failed setns to connector network namespace: %s\n", strerror(errno));
+
+		ret = close(sk_fds[0]);
+		if (ret < 0) {
+			fprintf(stderr, "%s - Failed to close socket fd %d\n",
+				strerror(errno), sk_fds[1]);
 			_exit(1);
 		}
+
+		ret = wait_for_pid(pid);
+		if (ret < 0) {
+			fprintf(stderr, "Failed to wait for pid %d\n", pid);
+			_exit(EXIT_FAILURE);
+		}
```

Comment:
Either the error string is wrong or we need an extra check.
We should be able to distinguish between waitpid() itself failing vs the child process having exited non-zero.

I think in this case, wait_for_pid does both and handles errors from waitpid internally, so the error string should be changed to say `Failed to start listener` or something along those lines.